File: ydb/library/yql/providers/generic/connector/app/server/utils/predicate_builder.go
package utils

import (
	"fmt"
	"strings"

	"github.com/ydb-platform/ydb-go-genproto/protos/Ydb"
	api_service_protos "github.com/ydb-platform/ydb/ydb/library/yql/providers/generic/connector/libgo/service/protos"
)

func formatValue(formatter SQLFormatter, args []any, value *Ydb.TypedValue) (string, []any, error) {
	switch v := value.Value.Value.(type) {
	case *Ydb.Value_BoolValue:
		return formatter.GetPlaceholder(len(args)), append(args, v.BoolValue), nil
	case *Ydb.Value_Int32Value:
		return formatter.GetPlaceholder(len(args)), append(args, v.Int32Value), nil
	case *Ydb.Value_Uint32Value:
		return formatter.GetPlaceholder(len(args)), append(args, v.Uint32Value), nil
	case *Ydb.Value_Int64Value:
		return formatter.GetPlaceholder(len(args)), append(args, v.Int64Value), nil
	case *Ydb.Value_Uint64Value:
		return formatter.GetPlaceholder(len(args)), append(args, v.Uint64Value), nil
	case *Ydb.Value_FloatValue:
		return formatter.GetPlaceholder(len(args)), append(args, v.FloatValue), nil
	case *Ydb.Value_DoubleValue:
		return formatter.GetPlaceholder(len(args)), append(args, v.DoubleValue), nil
	default:
		return "", args, fmt.Errorf("%w, type: %T", ErrUnimplementedTypedValue, v)
	}
}

func formatColumn(formatter SQLFormatter, args []any, col string) (string, []any, error) {
	return formatter.SanitiseIdentifier(col), args, nil
}

func formatNull(formatter SQLFormatter, args []any, n *api_service_protos.TExpression_TNull) (string, []any, error) {
	return "NULL", args, nil
}

func formatArithmeticalExpression(formatter SQLFormatter, args []any, expression *api_service_protos.TExpression_TArithmeticalExpression) (string, []any, error) {
	var operation string

	switch op := expression.Operation; op {
	case api_service_protos.TExpression_TArithmeticalExpression_MUL:
		operation = " * "
	case api_service_protos.TExpression_TArithmeticalExpression_ADD:
		operation = " + "
	case api_service_protos.TExpression_TArithmeticalExpression_SUB:
		operation = " - "
	case api_service_protos.TExpression_TArithmeticalExpression_BIT_AND:
		operation = " & "
	case api_service_protos.TExpression_TArithmeticalExpression_BIT_OR:
		operation = " | "
	case api_service_protos.TExpression_TArithmeticalExpression_BIT_XOR:
		operation = " ^ "
	default:
		return "", args, fmt.Errorf("%w, op: %d", ErrUnimplementedArithmeticalExpression, op)
	}

	left, args, err := formatExpression(formatter, args, expression.LeftValue)
	if err != nil {
		return "", args, fmt.Errorf("failed to format left argument: %w", err)
	}

	right, args, err := formatExpression(formatter, args, expression.RightValue)
	if err != nil {
		return "", args, fmt.Errorf("failed to format right argument: %w", err)
	}

	return fmt.Sprintf("(%s%s%s)", left, operation, right), args, nil
}

func formatExpression(formatter SQLFormatter, args []any, expression *api_service_protos.TExpression) (string, []any, error) {
	if !formatter.SupportsPushdownExpression(expression) {
		return "", args, ErrUnsupportedExpression
	}

	switch e := expression.Payload.(type) {
	case *api_service_protos.TExpression_Column:
		return formatColumn(formatter, args, e.Column)
	case *api_service_protos.TExpression_TypedValue:
		return formatValue(formatter, args, e.TypedValue)
	case *api_service_protos.TExpression_ArithmeticalExpression:
		return formatArithmeticalExpression(formatter, args, e.ArithmeticalExpression)
	case *api_service_protos.TExpression_Null:
		return formatNull(formatter, args, e.Null)
	default:
		return "", args, fmt.Errorf("%w, type: %T", ErrUnimplementedExpression, e)
	}
}

func formatComparison(formatter SQLFormatter, args []any, comparison *api_service_protos.TPredicate_TComparison) (string, []any, error) {
	var operation string

	switch op := comparison.Operation; op {
	case api_service_protos.TPredicate_TComparison_L:
		operation = " < "
	case api_service_protos.TPredicate_TComparison_LE:
		operation = " <= "
	case api_service_protos.TPredicate_TComparison_EQ:
		operation = " = "
	case api_service_protos.TPredicate_TComparison_NE:
		operation = " <> "
	case api_service_protos.TPredicate_TComparison_GE:
		operation = " >= "
	case api_service_protos.TPredicate_TComparison_G:
		operation = " > "
	default:
		return "", args, fmt.Errorf("%w, op: %d", ErrUnimplementedOperation, op)
	}

	left, args, err := formatExpression(formatter, args, comparison.LeftValue)
	if err != nil {
		return "", args, fmt.Errorf("failed to format left argument: %w", err)
	}

	right, args, err := formatExpression(formatter, args, comparison.RightValue)
	if err != nil {
		return "", args, fmt.Errorf("failed to format right argument: %w", err)
	}

	return fmt.Sprintf("(%s%s%s)", left, operation, right), args, nil
}

func formatNegation(formatter SQLFormatter, args []any, negation *api_service_protos.TPredicate_TNegation) (string, []any, error) {
	pred, args, err := formatPredicate(formatter, args, negation.Operand, false)
	if err != nil {
		return "", args, fmt.Errorf("failed to format NOT statement: %w", err)
	}

	return fmt.Sprintf("(NOT %s)", pred), args, nil
}

func formatConjunction(formatter SQLFormatter, args []any, conjunction *api_service_protos.TPredicate_TConjunction, topLevel bool) (string, []any, error) {
	var (
		sb        strings.Builder
		succeeded int32 = 0
		statement string
		err       error
		first     string
	)

	for _, predicate := range conjunction.Operands {
		argsCut := args
		statement, args, err = formatPredicate(formatter, args, predicate, false)

		if err != nil {
			if !topLevel {
				return "", args, fmt.Errorf("failed to format AND statement: %w", err)
			} else {
				args = argsCut
			}
		} else {
			if succeeded > 0 {
				if succeeded == 1 {
					sb.WriteString("(")
					sb.WriteString(first)
				}

				sb.WriteString(" AND ")
				sb.WriteString(statement)
			} else {
				first = statement
			}

			succeeded++
		}
	}

	if succeeded == 0 {
		return "", args, fmt.Errorf("failed to format AND statement: %w", err)
	}

	if succeeded == 1 {
		sb.WriteString(first)
	} else {
		sb.WriteString(")")
	}

	return sb.String(), args, nil
}

func formatDisjunction(formatter SQLFormatter, args []any, disjunction *api_service_protos.TPredicate_TDisjunction) (string, []any, error) {
	var (
		sb        strings.Builder
		cnt       int32 = 0
		statement string
		err       error
		first     string
	)

	for _, predicate := range disjunction.Operands {
		statement, args, err = formatPredicate(formatter, args, predicate, false)
		if err != nil {
			return "", args, fmt.Errorf("failed to format OR statement: %w", err)
		} else {
			if cnt > 0 {
				if cnt == 1 {
					sb.WriteString("(")
					sb.WriteString(first)
				}

				sb.WriteString(" OR ")
				sb.WriteString(statement)
			} else {
				first = statement
			}

			cnt++
		}
	}

	if cnt == 0 {
		return "", args, fmt.Errorf("failed to format OR statement: no operands")
	}

	if cnt == 1 {
		sb.WriteString(first)
	} else {
		sb.WriteString(")")
	}

	return sb.String(), args, nil
}

func formatIsNull(formatter SQLFormatter, args []any, isNull *api_service_protos.TPredicate_TIsNull) (string, []any, error) {
	statement, args, err := formatExpression(formatter, args, isNull.Value)
	if err != nil {
		return "", args, fmt.Errorf("failed to format IS NULL statement: %w", err)
	}

	return fmt.Sprintf("(%s IS NULL)", statement), args, nil
}

func formatIsNotNull(formatter SQLFormatter, args []any, isNotNull *api_service_protos.TPredicate_TIsNotNull) (string, []any, error) {
	statement, args, err := formatExpression(formatter, args, isNotNull.Value)
	if err != nil {
		return "", args, fmt.Errorf("failed to format IS NOT NULL statement: %w", err)
	}

	return fmt.Sprintf("(%s IS NOT NULL)", statement), args, nil
}

func formatPredicate(formatter SQLFormatter, args []any, predicate *api_service_protos.TPredicate, topLevel bool) (string, []any, error) {
	switch p := predicate.Payload.(type) {
	case *api_service_protos.TPredicate_Negation:
		return formatNegation(formatter, args, p.Negation)
	case *api_service_protos.TPredicate_Conjunction:
		return formatConjunction(formatter, args, p.Conjunction, topLevel)
	case *api_service_protos.TPredicate_Disjunction:
		return formatDisjunction(formatter, args, p.Disjunction)
	case *api_service_protos.TPredicate_IsNull:
		return formatIsNull(formatter, args, p.IsNull)
	case *api_service_protos.TPredicate_IsNotNull:
		return formatIsNotNull(formatter, args, p.IsNotNull)
	case *api_service_protos.TPredicate_Comparison:
		return formatComparison(formatter, args, p.Comparison)
	case *api_service_protos.TPredicate_BoolExpression:
		return formatExpression(formatter, args, p.BoolExpression.Value)
	default:
		return "", args, fmt.Errorf("%w, type: %T", ErrUnimplementedPredicateType, p)
	}
}

func formatWhereClause(formatter SQLFormatter, where *api_service_protos.TSelect_TWhere) (string, []any, error) {
	if where.FilterTyped == nil {
		return "", nil, ErrUnimplemented
	}

	args := make([]any, 0)
	formatted, args, err := formatPredicate(formatter, args, where.FilterTyped, true)

	if err != nil {
		return "", nil, err
	}

	result := "WHERE " + formatted

	return result, args, nil
}

File: ydb/core/fq/libs/row_dispatcher/predicate_builder.cpp
#include "predicate_builder.h"

#include <util/string/cast.h>

namespace NFq {

TString FormatColumn(const TString& value);
TString FormatValue(Ydb::TypedValue value) ;
TString FormatNull(NYql::NPq::NProto::TExpression_TNull);
TString FormatExpression(NYql::NPq::NProto::TExpression expression);
TString FormatArithmeticalExpression(NYql::NPq::NProto::TExpression_TArithmeticalExpression expression);
TString FormatNegation(const NYql::NPq::NProto::TPredicate_TNegation& negation);
TString FormatComparison(NYql::NPq::NProto::TPredicate_TComparison comparison);
TString FormatConjunction(const NYql::NPq::NProto::TPredicate_TConjunction& conjunction, bool topLevel);
TString FormatDisjunction(const NYql::NPq::NProto::TPredicate_TDisjunction& disjunction);
TString FormatIsNull(NYql::NPq::NProto::TPredicate_TIsNull isNull);
TString FormatIsNotNull(NYql::NPq::NProto::TPredicate_TIsNotNull isNotNull);
TString FormatPredicate(const NYql::NPq::NProto::TPredicate& predicate, bool topLevel);

TString FormatColumn(const TString& value) {
    return value;
}

TString FormatValue(Ydb::TypedValue value) {
	switch (value.value().value_case()) {
	case  Ydb::Value::kBoolValue:
		return ToString(value.value().bool_value());
	case Ydb::Value::kInt32Value:
		return ToString(value.value().int32_value());
	case Ydb::Value::kUint32Value:
		return ToString(value.value().uint32_value());
	case Ydb::Value::kInt64Value:
		return ToString(value.value().int64_value());
	case Ydb::Value::kUint64Value:
		return ToString(value.value().uint64_value());
	case Ydb::Value::kFloatValue:
		return ToString(value.value().float_value());
	case Ydb::Value::kDoubleValue:
		return ToString(value.value().double_value());
	case Ydb::Value::kBytesValue:
		return "\"" + ToString(value.value().bytes_value()) + "\"";
	case Ydb::Value::kTextValue:
		return "\"" + ToString(value.value().text_value()) + "\"";
	default:
        ythrow yexception() << "ErrUnimplementedTypedValue";
	}
}

TString FormatNull(NYql::NPq::NProto::TExpression_TNull) {
	return "NULL";
}

TString FormatExpression(NYql::NPq::NProto::TExpression expression) {
    switch (expression.payload_case()) {
	case NYql::NPq::NProto::TExpression::kColumn:
		return FormatColumn(expression.column());
	case NYql::NPq::NProto::TExpression::kTypedValue:
	 	return FormatValue(expression.typed_value());
    case NYql::NPq::NProto::TExpression::kArithmeticalExpression:
	  	return FormatArithmeticalExpression(expression.arithmetical_expression());
	case NYql::NPq::NProto::TExpression::kNull:
	 	return FormatNull(expression.null());
	default:
		ythrow yexception() << "UnimplementedExpression";
	}
}

TString FormatArithmeticalExpression(NYql::NPq::NProto::TExpression_TArithmeticalExpression expression) {
    TString operation;
	switch (expression.operation()) {
	case NYql::NPq::NProto::TExpression_TArithmeticalExpression::MUL:
		operation = " * ";
        break;
	case NYql::NPq::NProto::TExpression_TArithmeticalExpression::ADD:
		operation = " + ";
        break;
	case NYql::NPq::NProto::TExpression_TArithmeticalExpression::SUB:
		operation = " - ";
        break;
	case NYql::NPq::NProto::TExpression_TArithmeticalExpression::BIT_AND:
		operation = " & ";
        break;
	case NYql::NPq::NProto::TExpression_TArithmeticalExpression::BIT_OR:
		operation = " | ";
        break;
	case NYql::NPq::NProto::TExpression_TArithmeticalExpression::BIT_XOR:
		operation = " ^ ";
        break;
	default:
		ythrow yexception() << "ErrUnimplementedArithmeticalExpression";
	}

	auto left = FormatExpression(expression.left_value());
	auto right = FormatExpression(expression.right_value());
	return left + operation + right;
}

TString FormatNegation(const NYql::NPq::NProto::TPredicate_TNegation& negation) {
	auto pred = FormatPredicate(negation.operand(), false);
	return "(NOT " + pred + ")";
}

TString FormatConjunction(const NYql::NPq::NProto::TPredicate_TConjunction& conjunction, bool /*topLevel*/) {
	ui32 succeeded = 0;
	TStringStream stream;
	TString first;

	for (const auto& predicate : conjunction.operands()) {
		auto statement = FormatPredicate(predicate, false);

		if (succeeded > 0) {
			if (succeeded == 1) {
				stream << "(";
				stream << first;
			}
			stream << " AND ";
			stream << statement;
		} else {
			first = statement;
		}
		succeeded++;
	}

	if (succeeded == 0) {
		ythrow yexception() << "failed to format AND statement, no operands";
	}

	if (succeeded == 1) {
		stream << first;
	} else {
		stream << ")";
	}

	return stream.Str();
}

TString FormatDisjunction(const NYql::NPq::NProto::TPredicate_TDisjunction& disjunction) {
	TStringStream stream;
	TString first;
	ui32 cnt = 0;

	for (const auto& predicate : disjunction.operands()) {
		auto statement = FormatPredicate(predicate, false);

		if (cnt > 0) {
			if (cnt == 1) {
				stream << "(";
				stream << first;
			}

			stream << " OR ";
			stream << statement;
		} else {
			first = statement;
		}
		cnt++;
	}

	if (cnt == 0) {
		ythrow yexception() << "failed to format OR statement: no operands";
	}

	if (cnt == 1) {
		stream << first;
	} else {
		stream << ")";
	}

	return stream.Str();
}

TString FormatIsNull(NYql::NPq::NProto::TPredicate_TIsNull isNull) {
	auto statement = FormatExpression(isNull.value());
	return "(" + statement + " IS NULL)";
}

TString FormatIsNotNull(NYql::NPq::NProto::TPredicate_TIsNotNull isNotNull) {
	auto statement = FormatExpression(isNotNull.value());
	return "(" + statement + " IS NOT NULL)";
}

TString FormatComparison(NYql::NPq::NProto::TPredicate_TComparison comparison) {
	TString operation;

    switch (comparison.operation()) {
	case NYql::NPq::NProto::TPredicate_TComparison::L:
		operation = " < ";
        break;
	case NYql::NPq::NProto::TPredicate_TComparison::LE:
		operation = " <= ";
        break;
	case NYql::NPq::NProto::TPredicate_TComparison::EQ:
		operation = " = ";
        break;
	case NYql::NPq::NProto::TPredicate_TComparison::NE:
		operation = " <> ";
        break;
	case NYql::NPq::NProto::TPredicate_TComparison::GE:
		operation = " >= ";
        break;
	case NYql::NPq::NProto::TPredicate_TComparison::G:
		operation = " > ";
        break;
	default:
		ythrow yexception() << "UnimplementedOperation";
	}

	auto left = FormatExpression(comparison.left_value());
	auto right = FormatExpression(comparison.right_value());

	return left + operation + right;
}

TString FormatPredicate(const NYql::NPq::NProto::TPredicate& predicate, bool topLevel ) {
    switch (predicate.payload_case()) {
        case NYql::NPq::NProto::TPredicate::PAYLOAD_NOT_SET:
                return {};
        case NYql::NPq::NProto::TPredicate::kNegation:
    		return FormatNegation(predicate.negation());
        case NYql::NPq::NProto::TPredicate::kConjunction:
            return FormatConjunction(predicate.conjunction(), topLevel);
        case NYql::NPq::NProto::TPredicate::kDisjunction:
            return FormatDisjunction(predicate.disjunction());
        case NYql::NPq::NProto::TPredicate::kIsNull:
            return FormatIsNull(predicate.is_null());
        case NYql::NPq::NProto::TPredicate::kIsNotNull:
            return FormatIsNotNull(predicate.is_not_null());
        case NYql::NPq::NProto::TPredicate::kComparison:
            return FormatComparison(predicate.comparison());
        case NYql::NPq::NProto::TPredicate::kBoolExpression:
            return FormatExpression(predicate.bool_expression().value());
        default:
            ythrow yexception() << "UnimplementedPredicateType";
	}
}

TString FormatWhere(const NYql::NPq::NProto::TPredicate& predicate) {
    auto stream = FormatPredicate(predicate, true);
	if (stream.empty()) {
		return "";
	}
    return "WHERE " + stream;
}

} // namespace NFq
